param/__init__.py

```diff
@@ -1287,6 +1287,73 @@ def __init__(self, **params):
         super(Array,self).__init__(ndarray, allow_None=True, **params)
 
 
+class DataFrame(ClassSelector):
+    """
+    Parameter whose value is a pandas DataFrame.
+    """
+    __slots__ = ['rows','columns']
+
+    def __init__(self, rows=None, columns=None,**params):
+        from pandas import DataFrame as pdDFrame
+        self.rows = rows
+        self.columns = columns
+        super(DataFrame,self).__init__(pdDFrame, allow_None=True, **params)
+        self._check_value(self.default)
+
+    def _check_value(self,val,obj=None):
+        super(DataFrame, self)._check_value(val, obj)
+        length_error = 'Provided DataFrame has {found} columns when {expected} were expected'
+
+        if self.columns is None:
+            pass
+        elif isinstance(self.columns, list):
+            if len(val.columns) != len(self.columns):
+                raise Exception(length_error.format(found=len(val.columns), expected=len(self.columns)))
+            difference = set(val.columns) ^ set(self.columns)
+            if difference:
+                msg = 'Provided DataFrame columns {found} differ from expected columns {expected}'
+                raise Exception(msg.format(found=list(val.columns), expected=self.columns))
+        else:
+            if len(val.columns) != self.columns:
+                raise Exception(length_error.format(found=len(val.columns), expected=self.columns))
+
+        if self.rows is None:
+            pass
+        elif len(val) != self.rows:
+            msg = 'Provided DataFrame has {found} rows which does not match the expected {expected} rows'
+            raise Exception(msg.format(found=len(val), expected=self.rows))
```

Comment:
For rows, I can easily imagine wanting to express a minimum, maximum, or exact number, which if implemented for rows may as well be implemented for columns with the same syntax if it doesn't conflict with the specific list of columns support.  The current "exactly 3" approach for rows seems unlikely to be sufficient in general.